lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+    }
+
+    if (Allowed.empty()) {
+      // A return value is expected on an ABI that doesn't dedicate any
+      // registers to be used for those.
+      return std::nullopt;
     }
 
-    if (UsedRegisters.size() == 1) {
-      if (UsedRegisters.begin()->Location == PointerToCopyLocation) {
-        if constexpr (DryRun)
-          return model::QualifiedType{};
-        else
-          return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                  PointerToCopyLocation,
-                                  TheBinary);
+    // Special case where the only used register is the pointer-to-copy location
+    if (Used.size() == 1 && Used.begin()->Location == PointerToCopyLocation) {
+      if (Used.begin()->Type.isPointer()) {
+        // If the type is already a pointer, preserve it.
+        return Used.begin()->Type;
       } else {
-        if constexpr (RegisterCount == 0)
-          return std::nullopt;
-        if (AllowedRegisters.front() == UsedRegisters.begin()->Location) {
-          if constexpr (DryRun)
-            return model::QualifiedType{};
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
+    llvm::SmallVector<std::optional<model::QualifiedType>, 4> ResultingTypes;
+    bool MustUseTheNextOne = false;
+    auto AllowedRange = llvm::enumerate(llvm::reverse(Allowed));
+    for (auto Pair : AllowedRange) {
+      size_t Index = Allowed.size() - Pair.index() - 1;
+      model::Register::Values Register = Pair.value();
+      auto UsedIterator = Used.find(Register);
+      if (UsedIterator != Used.end())
+        ResultingTypes.emplace_back(UsedIterator->Type);
+      else if (MustUseTheNextOne)
+        ResultingTypes.emplace_back(std::nullopt);
+
+      MustUseTheNextOne = MustUseTheNextOne || (UsedIterator != Used.end());
+    }
```

Comment:
I'm a bit confused by this. I think we should:

1. Ensure that `Used` is a subset of `Allowed`, if not, bail out.
2. Ensure that if the second return register is used (e.g., `rdx`) then also the first is used (e.g., `rax`). Bail out in all other situations.

Right now it seems to me that we try to process the input even if `Used` contains stuff not contained in `Allowed`.

Reply:
I guess that's because this stuff was written before the `detect-abi` started filling in the missing registers so I hand-rolled a solution to work around that until it was ready, now this can, luckily, be retired, as long as we somehow ensure that `detect-abi` has been ran on any binary we try to push through this pass.

Reply:
Thanks to `RegisterStateDeductions`, it should now function the way you expect as I'm opting for strict enforcement everywhere where abi conversions are concerned.